In [1]:
import numpy as np

In [3]:
arr = np.arange(11,22)
arr

array([11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21])

In [4]:
#Bracket Indexing and Selection
arr[4]

15

In [5]:
#values in a range
arr[4:7]

array([15, 16, 17])

# Broadcasting

In [7]:
arr[0:5] =10

In [8]:
arr

array([10, 10, 10, 10, 10, 16, 17, 18, 19, 20, 21])

In [10]:
#reset the array
arr = np.arange(11,22)

In [11]:
slice_of_arr = arr[0:5]

In [12]:
slice_of_arr

array([11, 12, 13, 14, 15])

In [17]:
slice_of_arr[:] = 99

In [18]:
slice_of_arr

array([99, 99, 99, 99, 99])

In [19]:
arr

array([99, 99, 99, 99, 99, 16, 17, 18, 19, 20, 21])

In [20]:
slice_of_arr = arr[0:5].copy()

In [21]:
slice_of_arr[:] = 59

In [22]:
arr

array([99, 99, 99, 99, 99, 16, 17, 18, 19, 20, 21])

In [23]:
slice_of_arr

array([59, 59, 59, 59, 59])

### Indexing a 2D array

In [26]:
arr_2d = np.array(([5,10,15],[20,25,30],[35,40,45]))
arr_2d

array([[ 5, 10, 15],
       [20, 25, 30],
       [35, 40, 45]])

In [28]:
#Indexing the row
arr_2d[1]

array([20, 25, 30])

In [29]:
arr_2d[1][-2]

25

In [35]:
arr_2d[:][-2]

array([20, 25, 30])

In [36]:
#arr_2d[row,col]

arr_2d[-1,-1]

45

In [37]:
arr_2d[2,-1]

45

In [38]:
arr_2d[-1,2]

45

In [39]:
arr_2d[2,2]

45

In [40]:
arr_2d

array([[ 5, 10, 15],
       [20, 25, 30],
       [35, 40, 45]])

In [41]:
arr_2d[:2,:2]

array([[ 5, 10],
       [20, 25]])

In [42]:
arr_2d.diagonal()

array([ 5, 25, 45])

In [43]:
arr_2d[1:,1:]

array([[25, 30],
       [40, 45]])

In [48]:
arr_eg = np.linspace(0,5,35).reshape(7,5)
arr_eg

array([[0.        , 0.14705882, 0.29411765, 0.44117647, 0.58823529],
       [0.73529412, 0.88235294, 1.02941176, 1.17647059, 1.32352941],
       [1.47058824, 1.61764706, 1.76470588, 1.91176471, 2.05882353],
       [2.20588235, 2.35294118, 2.5       , 2.64705882, 2.79411765],
       [2.94117647, 3.08823529, 3.23529412, 3.38235294, 3.52941176],
       [3.67647059, 3.82352941, 3.97058824, 4.11764706, 4.26470588],
       [4.41176471, 4.55882353, 4.70588235, 4.85294118, 5.        ]])

* First 3 rows and all columns
* rows from 0:6 and all columns
* array of shape 1*9 with elements: 

* 3.08823529, 3.23529412, 3.38235294,3.82352941, 3.97058824, 4.11764706,4.55882353, 4.70588235, 4.85294118

In [50]:
arr_eg[:3]

array([[0.        , 0.14705882, 0.29411765, 0.44117647, 0.58823529],
       [0.73529412, 0.88235294, 1.02941176, 1.17647059, 1.32352941],
       [1.47058824, 1.61764706, 1.76470588, 1.91176471, 2.05882353]])

In [60]:
arr_eg

array([[0.        , 0.14705882, 0.29411765, 0.44117647, 0.58823529],
       [0.73529412, 0.88235294, 1.02941176, 1.17647059, 1.32352941],
       [1.47058824, 1.61764706, 1.76470588, 1.91176471, 2.05882353],
       [2.20588235, 2.35294118, 2.5       , 2.64705882, 2.79411765],
       [2.94117647, 3.08823529, 3.23529412, 3.38235294, 3.52941176],
       [3.67647059, 3.82352941, 3.97058824, 4.11764706, 4.26470588],
       [4.41176471, 4.55882353, 4.70588235, 4.85294118, 5.        ]])

In [59]:
arr_eg[4:,1:4].reshape(1,9)

array([[3.08823529, 3.23529412, 3.38235294, 3.82352941, 3.97058824,
        4.11764706, 4.55882353, 4.70588235, 4.85294118]])

In [63]:
arr_eg[-3:,-4:-1].reshape(1,9)

array([[3.08823529, 3.23529412, 3.38235294, 3.82352941, 3.97058824,
        4.11764706, 4.55882353, 4.70588235, 4.85294118]])

In [65]:
arr_eg[-3:,:2]

array([[2.94117647, 3.08823529],
       [3.67647059, 3.82352941],
       [4.41176471, 4.55882353]])

In [67]:
arr_eg[-4:,:2]

array([[2.20588235, 2.35294118],
       [2.94117647, 3.08823529],
       [3.67647059, 3.82352941],
       [4.41176471, 4.55882353]])

## Conditional selection

In [68]:
arr2 = np.arange(0,11)

In [69]:
arr2

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10])

In [70]:
arr2 > 4

array([False, False, False, False, False,  True,  True,  True,  True,
        True,  True])

In [71]:
arr2[arr2 > 4]

array([ 5,  6,  7,  8,  9, 10])

In [76]:
arr_eg[:,-2] > 2.5

array([False, False, False,  True,  True,  True,  True])

In [74]:
arr[arr_eg > 2.5]

IndexError: too many indices for array: array is 1-dimensional, but 2 were indexed

In [77]:
arr_eg[:,-2][arr_eg[:,-2] > 2.5]

array([2.64705882, 3.38235294, 4.11764706, 4.85294118])

In [78]:
arr_eg

array([[0.        , 0.14705882, 0.29411765, 0.44117647, 0.58823529],
       [0.73529412, 0.88235294, 1.02941176, 1.17647059, 1.32352941],
       [1.47058824, 1.61764706, 1.76470588, 1.91176471, 2.05882353],
       [2.20588235, 2.35294118, 2.5       , 2.64705882, 2.79411765],
       [2.94117647, 3.08823529, 3.23529412, 3.38235294, 3.52941176],
       [3.67647059, 3.82352941, 3.97058824, 4.11764706, 4.26470588],
       [4.41176471, 4.55882353, 4.70588235, 4.85294118, 5.        ]])

In [84]:
arr_eg.reshape(35,)[arr_eg.reshape(35,) > 2].reshape(7,3)

array([[2.05882353, 2.20588235, 2.35294118],
       [2.5       , 2.64705882, 2.79411765],
       [2.94117647, 3.08823529, 3.23529412],
       [3.38235294, 3.52941176, 3.67647059],
       [3.82352941, 3.97058824, 4.11764706],
       [4.26470588, 4.41176471, 4.55882353],
       [4.70588235, 4.85294118, 5.        ]])

In [86]:
arr_eg_1d = arr_eg.reshape(35,)
arr_eg_1d

array([0.        , 0.14705882, 0.29411765, 0.44117647, 0.58823529,
       0.73529412, 0.88235294, 1.02941176, 1.17647059, 1.32352941,
       1.47058824, 1.61764706, 1.76470588, 1.91176471, 2.05882353,
       2.20588235, 2.35294118, 2.5       , 2.64705882, 2.79411765,
       2.94117647, 3.08823529, 3.23529412, 3.38235294, 3.52941176,
       3.67647059, 3.82352941, 3.97058824, 4.11764706, 4.26470588,
       4.41176471, 4.55882353, 4.70588235, 4.85294118, 5.        ])

In [87]:
arr_eg_1d[arr_eg_1d > 2].reshape(7,3)

array([[2.05882353, 2.20588235, 2.35294118],
       [2.5       , 2.64705882, 2.79411765],
       [2.94117647, 3.08823529, 3.23529412],
       [3.38235294, 3.52941176, 3.67647059],
       [3.82352941, 3.97058824, 4.11764706],
       [4.26470588, 4.41176471, 4.55882353],
       [4.70588235, 4.85294118, 5.        ]])